### ISSUE TITLE
bill of materials output for easy optimization?

### ISSUE THREAD

**Issue by timvisher** [overviewer/Minecraft-Overviewer#1242]
__tl;dr:__

Is it possible for OV to generate a bill of materials that it generated so that other tools can work on it?

__context:__

I'm trying to write a build system to update my overviewer site and my efforts are being thwarted. :)

Essentially, I'm serving the overviewer out of s3 and I'm attempting to save cost and bandwidth by gzipping and optipnging the content.

My initial thought with this was to capture the start time of a given run, and then run something along the lines of

```
find mc-overviewer-maps \
     -type f \
     -newer overviewer-start-time \
     -print \
     > overviewer-newer-files
```

To get the files that had been generated by the most recent run, and then only copy/optipng/gzip them.

That strategy seems to be bust, unfortunately. It seems to work most of the time, but it appears that the file's mtime's are not always set to what I would expect.

For instance, in the most recent run (at around 6am this morning, EST) a new file was generated with an mtime in the past (unless I'm misunderstanding what `ls -lT` outputs).

```
🐉  ls -lT mc-overviewer-maps/survivalday/3/1/0/0/1/0.png
-rw-r--r--  1 ---  ---  31065 Jul 11 21:38:11 2015 ---/mc-overviewer-maps/survivalday/3/1/0/0/1/0.png
```

I'm very confused what that time could possibly be, actually, given that at least some of the files generated have the expected mtimes that are after the start time.

Do you have any feedback on the general strategy here?

**Comment by timvisher** [comment]
It did just occur to me that it might be that the mtimes are being set to the time when the chunk was actually last modified, which would mean that I could optimized based on the last time the world was backed up (since I'm generating the map based on backup snapshots).

**Comment by CounterPillow** [comment]
First of all

>Essentially, I'm serving the overviewer out of s3 and I'm attempting to save cost and bandwidth by gzipping and optipnging the content.

gzipping is useless in this case. PNG already uses DEFLATE internally. You're better off using pngnq + optipng. Overviewer can do this for you using `optimizeimg`. If you live in a world with infinite computer time, I could also add zopflipng support to it, as zopflipng usually shaves off another 5%.

>For instance, in the most recent run (at around 6am this morning, EST) a new file was generated with an mtime in the past (unless I'm misunderstanding what ls -lT outputs).

Judging from #1241 this could be in relation to your setup, but that's just a guess.

**Comment by timvisher** [comment]
> Judging from #1241 this could be in relation to your setup, but that's just a guess.

So you would expect that the files generated by OV would have mtimes later than the start of the run?

**Comment by timvisher** [comment]
> First of all
> 
> > Essentially, I'm serving the overviewer out of s3 and I'm attempting to save cost and bandwidth by gzipping and optipnging the content.
> 
> gzipping is useless in this case. PNG already uses DEFLATE internally. You're better off using pngnq + optipng. Overviewer can do this for you using optimizeimg. If you live in a world with infinite computer time, I could also add zopflipng support to it, as zopflipng usually shaves off another 5%.

Ah! That's great to know. I knew that optipng was helping quite a bit but I don't think I checked the gzip savings. I hadn't heard of pngnq. I'm also trying to simplify the build process and not set content-encodings differently on a per file basis. I'll check into the optimizeimg flag as well. At least that would remove one build step.

**Comment by CounterPillow** [comment]
>I hadn't heard of pngnq. I'm also trying to simplify the build process and not set content-encodings differently on a per file basis. I'll check into the optimizeimg flag as well. At least that would remove one build step.

See http://overviewer.org/blog/2014/5/9/new-optimizeimg/

>but I don't think I checked the gzip savings.

They're likely not savings at all.

The gzip metadata is like 18 bytes or something, unless you manage to compress the ~64 bytes of PNG metadata by more than 18 bytes it actually adds to the filesize, and that's ignoring what gzip does with the already zlib compressed PNG IDAT data.

>So you would expect that the files generated by OV would have mtimes later than the start of the run?

Yes. Overviewer's tile mtimes are assigned by the operating system, from what I can tell.

**Comment by agrif** [comment]
Hello! The optimizeimg options ought to do the optimizations you want, and if there's some other post-processing command you wish to run, it can probably be plugged in to the optimizeimg framework.

As for mtimes, Overviewer definitely sets them to the most recent modification time of all the chunks each tile contains. If you want to figure out what files are modified by overviewer, you could read in the last render time written out to `overviewerConfig.js` before running a new render, then look for all tiles with an mtime later than that. This is the same method Overviewer uses to decide what new tiles to render, based on chunk mtimes.

**Comment by CounterPillow** [comment]
Alternatively we also already have a changelist feature, though I'm not sure if it still works:

http://docs.overviewer.org/en/latest/config/?highlight=changelist#map-update-behavior

**Comment by timvisher** [comment]
@agrif Just to be sure I'm understanding you, the mtimes are the mtimes of the chunk that each tile contains, so it would make perfect sense that the mtime of a chunk that was modified and then unloaded prior to the run of OV to be in the past. But it should always be after the previous run started. That's the rule I'm going forward with (I timestamp the backups that I'm taking so I'm just parsing that and using it as input into the `find` command).

gzip is definitely not saving me any space. That's good to know.

**Comment by CounterPillow** [comment]
I'm going to close this for now since the feature you were asking for is already implemented (`changelist`), and a second feature is there too for your specific use-case (`optimizeimg`). If you have any further questions or comments regarding this issue, feel free to re-open.

**Comment by agrif** [comment]
Yes. After a render, any updated tiles will have an mtime that lies between the end of that render, and the beginning of the last render. It's probably a little bit tighter than that, but I can't say for sure, and hopefully the render times aren't so long that it matters a lot.

**Comment by timvisher** [comment]
@CounterPillow Makes sense. Is there a better forum to ask these more open-ended questions? I don't mind using Issues for this but it seems like a hammer for a screw. :)

**Comment by CounterPillow** [comment]
@timvisher We're usually around in our [IRC Channel on FreeNode](http://overviewer.org/irc/), feel free to drop by if you have any questions or just want to chat.
